builders/txDataBuilder.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		builder.err = err
+		return builder
+	}
+
+	builder.address = address.AddressAsBech32String()
+
+	return builder
+}
+
+// ArgHexString adds the provided hex string to the arguments list
+func (builder *txDataBuilder) ArgHexString(hexed string) TxDataBuilder {
+	_, err := hex.DecodeString(hexed)
+	if err != nil {
+		builder.err = fmt.Errorf("%w in builder.ArgHexString for string %s", err, hexed)
+		return builder
+	}
+
+	builder.args = append(builder.args, hexed)
+
+	return builder
+}
+
+// ArgAddress adds the provided address to the arguments list
+func (builder *txDataBuilder) ArgAddress(address core.AddressHandler) TxDataBuilder {
+	err := builder.checkAddress(address)
+	if err != nil {
+		builder.err = err
+		return builder
+	}
+
+	return builder.addBytes(address.AddressBytes())
+}
+
+func (builder *txDataBuilder) checkAddress(address core.AddressHandler) error {
+	if check.IfNil(address) {
+		return fmt.Errorf("%w in builder.checkAddress", ErrNilAddress)
+	}
+	if len(address.AddressBytes()) == 0 {
+		return fmt.Errorf("%w in builder.checkAddress", ErrInvalidAddress)
+	}
+
+	return nil
+}
+
+// ArgBigInt adds the provided value to the arguments list
+func (builder *txDataBuilder) ArgBigInt(value *big.Int) TxDataBuilder {
+	if value == nil {
+		builder.err = fmt.Errorf("%w in builder.ArgBigInt", ErrNilValue)
+		return builder
+	}
+
+	return builder.addBytes(value.Bytes())
+}
+
+// ArgInt64 adds the provided value to the arguments list
+func (builder *txDataBuilder) ArgInt64(value int64) TxDataBuilder {
+	b := big.NewInt(value)
+
+	return builder.addBytes(b.Bytes())
+}
+
+// ArgBytes adds the provided bytes to the arguments list. The parameter should contain at least one byte
+func (builder *txDataBuilder) ArgBytes(bytes []byte) TxDataBuilder {
+	if len(bytes) == 0 {
+		builder.err = fmt.Errorf("%w in builder.ArgBytes", ErrInvalidValue)
+	}
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
+
+	return builder
+}
+
+func (builder *txDataBuilder) addBytes(bytes []byte) TxDataBuilder {
+	if len(bytes) == 0 {
+		bytes = []byte{0}
+	}
+
+	builder.args = append(builder.args, hex.EncodeToString(bytes))
+
+	return builder
+}
+
+// ToDataString returns the formatted data string ready to be used in a transaction call
+func (builder *txDataBuilder) ToDataString() (string, error) {
+	if builder.err != nil {
+		return "", builder.err
+	}
+
+	parts := append([]string{builder.function}, builder.args...)
+
+	return strings.Join(parts, dataSeparator), nil
+}
+
+// ToDataBytes returns the formatted data string ready to be used in a transaction call as bytes
+func (builder *txDataBuilder) ToDataBytes() ([]byte, error) {
+	dataField, err := builder.ToDataString()
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(dataField), err
+}
+
+// ToVmValueRequest returns the VmValueRequest structure to be used in a VM call
+func (builder *txDataBuilder) ToVmValueRequest() (*data.VmValueRequest, error) {
```

Comment:
this is confusing. Why don't you create 2 different components? one for tx payload construction and one for vm values request builder.. or at least have 2 interfaces that are implemented by the same component (you'll have to return  `*txDataBuilder` instead of an interface. This way you'll have `TxPayloadBuilder` and `VmValuesQueryBuilder` which only expose functions correlated with their names.

If I only use your struct for preparing sc calls, seeing `ToVmValueRequest()` or `CallerAddress` might not be the base case.

Reply:
split in 2. Extracted common code as maximum it can get.